pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go

```diff
@@ -65,6 +61,10 @@ func (w *Watcher) Start(stopCh <-chan struct{}) error {
 	if err != nil {
 		return fmt.Errorf("failed to start plugin fsWatcher, err: %v", err)
 	}
+	err = fsWatcher.Add(w.path)
+	if err != nil {
+		return fmt.Errorf("failed to watch %s, err: %v", w.path, err)
+	}
```

Comment:
it seems like the fundamental issue is a race inside traversePluginDir in the following scenario:

1. traversePluginDir("/foo") is called
2. `w.fs.Walk` enumerates children of `/foo`
3. a new child `/foo/bar` is created
4. the walker starts watching `/foo`

It seems like traversePluginDir should hoist watching dir before doing `w.fs.Walk`:

```
func (w *Watcher) traversePluginDir(dir string) error {
	// watch the new dir
	if err := w.fsWatcher.Add(path); err != nil {
		return fmt.Errorf("failed to watch %s, err: %v", path, err)
	}
	// traverse existing children in the dir
	return w.fs.Walk(dir, func(path string, info os.FileInfo, err error) error {
		...
	})
```

Reply:
if we fix traversePluginDir, I think we can drop this special-case call to `fsWatcher.Add(w.path)`

Reply:
Great! I think this is much better!